File: AgoraChat/Code/Controllers/Groups/ACDJoinGroupViewController.m
//
//  ACDJoinGroupViewController.m
//  ChatDemo-UI3.0
//
//  Created by liang on 2021/10/25.
//  Copyright © 2021 easemob. All rights reserved.
//

#import "ACDJoinGroupViewController.h"
#import "ACDSearchResultView.h"
#import "ACDSearchTableViewController.h"
#import "AgoraRealtimeSearchUtils.h"
#import "ACDSearchJoinCell.h"

#define kSearchBarHeight 40.0f

@interface ACDJoinGroupViewController ()
@property (nonatomic, strong) ACDSearchResultView *resultView;

@end

@implementation ACDJoinGroupViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    if (self.isSearchGroup) {
        self.title = @"Join a Group";
    }else {
        self.title = @"Add Contacts";
    }
    
    [self setNavBar];
    [self.table registerClass:[ACDSearchJoinCell class] forCellReuseIdentifier:[ACDSearchJoinCell reuseIdentifier]];
}

- (void)setNavBar {
    
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:ImageWithName(@"gray_goBack") style:UIBarButtonItemStylePlain target:self action:@selector(goBack)];
    
    UIButton *button = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 8, 15)];
    [button addTarget:self action:@selector(goBack) forControlEvents:UIControlEventTouchUpInside];
    
    [button setTitle:@"Cancel" forState:UIControlStateNormal];
    [button setTitleColor:TextLabelBlueColor forState:UIControlStateNormal];
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:button];
}

- (void)goBack {
    [self.navigationController popViewControllerAnimated:YES];
}


- (void)joinToPublicGroup:(NSString *)groupId {
    ACD_WS
    [MBProgressHUD showHUDAddedTo:[UIApplication sharedApplication].keyWindow
                         animated:YES];
    
    [[AgoraChatClient sharedClient].groupManager requestToJoinPublicGroup:groupId message:@"" completion:^(AgoraChatGroup *aGroup, AgoraChatError *aError) {
        [MBProgressHUD hideAllHUDsForView:[UIApplication sharedApplication].keyWindow animated:YES];
        if (!aError) {

        }
        else {
            NSString *msg = aError.errorDescription;
            [self showHint:msg];
        }
    }];
}


- (void)sendAddContact:(NSString *)contactName {
    NSString *requestMessage = [NSString stringWithFormat:@"%@ add you as a friend",AgoraChatClient.sharedClient.currentUsername];
    WEAK_SELF
    [MBProgressHUD showHUDAddedTo:self.navigationController.view animated:YES];
    [[AgoraChatClient sharedClient].contactManager addContact:contactName
                                               message:requestMessage
                                            completion:^(NSString *aUsername, AgoraChatError *aError) {
        [MBProgressHUD hideHUDForView:weakSelf.navigationController.view animated:YES];
        if (!aError) {
            NSString *msg =  @"You request has been sent.";
            [weakSelf showAlertWithMessage:msg];
        }
        else {
            [weakSelf showAlertWithMessage:aError.errorDescription];
        }
    }];
}



#pragma mark - UISearchBarDelegate
- (void)searchBar:(UISearchBar *)searchBar textDidChange:(NSString *)searchText {
    self.dataArray = [@[searchText] mutableCopy];
    self.searchSource = self.dataArray;
    self.table.userInteractionEnabled = YES;
    [self.table reloadData];
}

- (void)searchBarCancelButtonClicked:(UISearchBar *)searchBar {
    searchBar.text = @"";
    [searchBar setShowsCancelButton:NO animated:NO];
    [searchBar resignFirstResponder];
    self.searchSource = [@[] mutableCopy];
    self.table.scrollEnabled = NO;
    [self.table reloadData];
}



#pragma mark - Table view data source
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 54.0f;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.searchSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    ACDSearchJoinCell *cell = [tableView dequeueReusableCellWithIdentifier:[ACDSearchJoinCell reuseIdentifier]];
    if (cell == nil) {
        cell = [[ACDSearchJoinCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:[ACDSearchJoinCell reuseIdentifier]];
    }
    cell.isSearchGroup = self.isSearchGroup;
    
    NSString *title = @"";
    if (self.isSearchGroup) {
        title = [NSString stringWithFormat:@"GroupID：%@",self.searchSource[0]];
    }else {
        title = [NSString stringWithFormat:@"AgoraID：%@",self.searchSource[0]];
    }

    ACD_WS
    cell.addGroupBlock = ^{
        if (weakSelf.isSearchGroup) {
            [weakSelf joinToPublicGroup:weakSelf.searchSource[0]];

        }else {
            [weakSelf sendAddContact:weakSelf.searchSource[0]];
        }

    };

    [cell updateSearchName:title];
    return cell;
    
}


#pragma mark getter and setter
- (ACDSearchResultView *)resultView {
    if (_resultView == nil) {
        _resultView = [[ACDSearchResultView alloc] initWithFrame:CGRectMake(0, 0, KScreenWidth, 30.0)];
    }
    
    return _resultView;
}


@end

#undef kSearchBarHeight

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Groups/ACDJoinGroupViewController.swift
//
//  ACDJoinGroupViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/21.
//

import UIKit

class ACDJoinGroupViewController: ACDSearchTableViewController<String> {

    var isSearchGroup: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        if self.isSearchGroup {
            self.title = "Join a Group".localized
        } else {
            self.title = "Add Contacts".localized
        }
        
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "gray_goBack"), style: .plain, target: self, action: #selector(backAction))

        let btn = UIButton(frame: CGRect(x: 0, y: 0, width: 8, height: 15))
        btn.addTarget(self, action: #selector(backAction), for: .touchUpInside)
        btn.setTitle(LocalizedString.Cancel, for: .normal)
        btn.setTitleColor(Color_154DFE, for: .normal)
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: btn)
        
        self.table.rowHeight = 54
        self.table.register(UINib(nibName: "ACDSearchJoinCell", bundle: nil), forCellReuseIdentifier: "cell")
        
        self.searchBar.snp.remakeConstraints { make in
            make.top.equalTo(self.navigationController?.navigationBar.frame.height ?? 0)
            make.left.right.equalTo(self.view)
        }
    }

    @objc private func backAction() {
        self.navigationController?.popViewController(animated: true)
    }

    private func joinToPublicGroup(groupId: String) {
        self.cancelSearchState()
        MBProgressHUD.showAdded(to: self.view, animated: true)
        AgoraChatClient.shared().groupManager?.request(toJoinPublicGroup: groupId, message: "", completion: { group, error in
            MBProgressHUD.hide(for: self.view, animated: true)
            if let error = error {
                self.showHint(error.errorDescription)
            }
        })
    }
    
    private func sendAddContact(to contact: String) {
        if contact.count <= 0 {
            self.showHint(String(format: "%@ can't be null".localized, "Contact Name"))
            return
        }
        if self.isContainInMyContacts(contact) {
            self.searchBar.text = ""
            self.showHint("This contact has been added".localized)
            return
        }
        if contact.uppercased() == AgoraChatClient.shared().currentUsername?.uppercased() {
            self.searchBar.text = ""
            self.showHint("Not allowed to send their own friends to apply for".localized)
            return
        }
        self.cancelSearchState()
        
        let msg = String(format: "%@ add you as a friend".localized, AgoraChatClient.shared().currentUsername!)
        MBProgressHUD.showAdded(to: self.view, animated: true)
        AgoraChatClient.shared().contactManager?.addContact(contact, message: msg, completion: { username, error in
            MBProgressHUD.hide(for: self.view, animated: true)
            if let error = error {
                self.showAlert(message: error.errorDescription)
            } else {
                self.showAlert(message: "Operation succeeded".localized)
            }
        })
    }
    
    private func isContainInMyContacts(_ contact: String) -> Bool {
        if let contacts = AgoraChatClient.shared().contactManager?.getContacts() {
            return contacts.contains(contact)
        }
        return false
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.searchSource.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath)
        if let cell = cell as? ACDSearchJoinCell, let text = self.searchSource[0] as? String {
            cell.isSearchGroup = self.isSearchGroup
            cell.searchName = self.isSearchGroup ? "\("GroupID".localized)：\(text)" : "\("AgoraID".localized)：\(text)"
            cell.addGroupHandle = { [unowned self] in
                if self.isSearchGroup {
                    self.joinToPublicGroup(groupId: text)
                } else {
                    self.sendAddContact(to: text)
                }
            }
        }
        return cell
    }
    
    override func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
        self.dataArray = [searchText]
        self.searchSource = self.dataArray
        self.table.reloadData()
    }
    
    override func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
        searchBar.text = ""
        searchBar.setShowsCancelButton(false, animated: true)
        searchBar.resignFirstResponder()
        self.searchSource = []
        self.table.reloadData()
    }
}
